### DISCUSSION TITLE
C# proposal: implicitly convert nullable bool to bool

### DISCUSSION THREAD

**Discussion by mariusGundersen** [dotnet/csharplang#2861] [Language Ideas]
Given the following

```C#
bool? maybe = true;

if(maybe){
  Console.WriteLine("Success");
}
```

What is the expected result? The actual result is that compilation fails with the error "Cannot implicitly convert type 'bool?' to 'bool'.". That seems weird to me. I can understand why other nullable types cannot be implicitly converted to their non-nullable equivalent, but for `Nullable<bool>` that is slightly annoying, since now I need to write `if(maybe == true)`. 

I therefore propose that nullable bool can be implicitly converted to bool, such that `null` => `false`.

**Comment by stepanbenes** [comment]
The `null` value is generally considered as "I don't know" value. So, why should it be interpreted as `false`?

**Comment by spydacarnage** [comment]
If you consider that `if (maybe)` is shorthand for `if (maybe == true)`, yet they work differently.

If a `bool?` is null, then it isn't `true`, so while `null` shouldn't be interpreted as `false`, it should be interpreted as not `true`, and thus should fail `if (maybe)`, as opposed to having a compiler error.

**Reply by darraghjones** [reply]
I agree that it's confusing that `if (maybe) ` behaves differnetly to `if (maybe == true)`

Howerver, I would argue that implicitly converting a bool to a bool? should at least generate a warning.

**Reply by Joe4evr** [reply]
> Howerver, I would argue that implicitly converting a bool to a bool? should at least generate a warning.

You can always write your own analyzer to enforce that, but for the general language that ship has sailed when `Nullable<T>` was introduced.

**Reply by jnm2** [reply]
It's also confusing that `if (maybe)` behaves differently than `if (maybe != false)`, but that's not a reason to make them the same. You'd have to arbitrarily pick one consistency over the other.

**Comment by HaloFour** [comment]
This was an intentional design choice to force the developer to consider `bool?` as a proper tristate.  `null` is neither `true` nor `false` and treating it as either implicitly can lead to accidental logic errors.

**Reply by JustNrik** [reply]
This is a fallacy.

You're right, `null` is neither `true` nor `false`, and? what do you want to prove with it? 

`if` evaluates if the given expression returns `true`, nobody is implying that `null is false`, so this "intentional design choice" was based on a fallacy. `else` statement is not evaluate if the value is `false`, it's the fallback clause when the expression **is not true** which is perfectly valid for nullable booleans. This doesn't, by any means, imply that `null` is `false`. So please, stop hidding into a Vacuous truth fallacy. This was a bad design choice, period.

**Reply by The-Futurist** [reply]
@JustNrik - I disagree, the else clause is not a "fallback" at all. False is the opposite of True. 

I do not see why `bool?` should be treated as if it were in fact `bool`, if a `bool `is required then a `bool `should be used. Why use `bool?` if one does not want a simple true/false set of possible values?

And why assume that `null `be equivalent to `false`? there may be some problems where it makes sense to have `null `be equivalent to `true`, it depends on the problem surely?

Look at this:
```cs

    class Program
    {
        static void Main(string[] args)
        {
            Bool? bool_value = Bool.False;

            Trinary? trinary_value = Trinary.B;
        }
    }

    public enum Bool
    {
        False,
        True
    }

    public enum Trinary
    {
        A,
        B,
        C
    }
```

What if a Trinary value was in fact `null`? how would that be treated? as A? or B? or perhaps C?

The suggestion is akin to the sloppy semantics of C or C++, where a "true" is inferred for any value that has at least one non-zero bit in it, this has led to all sorts of bugs, it is sloppy to pretend a `bool?` can be used in an expression exactly as it were a `bool`.

It might be better in a real problem were one is dealing with `bool?` a lot, to define some static helper like False or True that accept a `bool?` and return whatever is useful for the problem being solved.

**Reply by JustNrik** [reply]
> 
> 
> @JustNrik - I disagree, the else clause is not a "fallback" at all. False is the opposite of True.

False is the opposite of true, yes, so? `else` doesn't mean "run this if the if expression is false", it means "run this if the if expression is not true", `not true` doesn't necessarily mean false, since `null` is also not true, and it's also not false, it's neither of them.

> I do not see why `bool?` should be treated as if it were in fact `bool`, if a `bool `is required then a `bool `should be used. Why use `bool?` if one does not want a simple true/false set of possible values?

Because sometimes we have an "optional" bool property which in most of cases we only care if it's true or not. And I don't really think it's wrong to assume that a `bool?` is either `true` or not since we shouldn't care about it being `false` or `null` at least when evaluating for thuthy.

> And why assume that `null `be equivalent to `false`? there may be some problems where it makes sense to have `null` be equivalent to `true`, it depends on the problem surely?

No one is assuming that "null is false", again, this is a fallacy. `null` is **neither** `true` nor `false`, and since it's **not true** it should fall into the else clause.

> Look at this:
> 
> ```cs
>     class Program
>     {
>         static void Main(string[] args)
>         {
>             Bool? bool_value = Bool.False;
> 
>             Trinary? trinary_value = Trinary.B;
>         }
>     }
> 
>     public enum Bool
>     {
>         False,
>         True
>     }
> 
>     public enum Trinary
>     {
>         A,
>         B,
>         C
>     }
> ```
> 
> What if a Trinary value was in fact `null`? how would that be treated? as A? or B? or perhaps C?

It should be threated as `null`, I don't know what are you trying to prove here.

> The suggestion is akin to the sloppy semantics of C or C++, where a "true" is inferred for any value that has at least one non-zero bit in it, this has led to all sorts of bugs, it is sloppy to pretend a `bool?` can be used in an expression exactly as it were a `bool`.

Not exactly, some things could legit be disallowed, like `if (!bool?)` since that would actually is ambiguous.
 
> It might be better in a real problem were one is dealing with `bool?` a lot, to define some static helper like False or True that accept a `bool?` and return whatever is useful for the problem being solved.

Perhaps.

**Reply by elpoul** [reply]
> False is the opposite of true, yes, so? `else` doesn't mean "run this if the if expression is false", it means "run this if the if expression is not true", `not true` doesn't necessarily mean false, since `null` is also not true, and it's also not false, it's neither of them.

You say that it doesn't mean "run this if expression is false" and yet this is exactly how C# spec defines an if statement:
```
if-statement:
  if ( boolean-expression ) embedded-statement
  if ( boolean-expression ) embedded-statement else embedded-statement
```
An if statement is executed as follows:
- The boolean-expression is evaluated
- If the Boolean expression yields true, control is transferred to the first embedded statement. When and if control reaches the end point of that statement, control is transferred to the end point of the if statement.
- If the Boolean expression yields false and if an else part is present, control is transferred to the second embedded statement. When and if control reaches the end point of that statement, control is transferred to the end point of the if statement

A nullable boolean expression is obviously not a boolean expression and since you have  3 possible states why not go for clarity and use either a switch statement (if all 3 states can lead to different outcomes) or a more explicit expression, like expr == true or expr ?? true, etc as many people have already suggested.

**Reply by The-Futurist** [reply]
Yes false _is_ the opposite of true, all boolean valued expressions are either true or false, in logic a proposition can be either true or false.

Because the expressions in a conditional branch must be either true or false then the _else_ is the opposite of the _then_.

So `not true` does actually mean false, if a proposition is not true it must be false else it would have to be true!

The OP suggested this:

`I therefore propose that nullable bool can be implicitly converted to bool, such that null => false.`

But why? why not suggest this:

`I therefore propose that nullable bool can be implicitly converted to bool, such that null => true.`

How do you decide which of these is preferable? IMHO this depends upon the problem itself so a fixed rule is inappropriate.

**Comment by elpoul** [comment]
`if (<expression>)` is a shorthand for `if (<expression> == true)` only if `<expression>` is a boolean expression itself. In my experience I've seen as many instances of `if (maybe ?? true)` as `if (maybe ?? false)` so I do believe that in nullable booleans it makes more sense to be explicit.

**Comment by canton7** [comment]
> If you consider that if (maybe) is shorthand for if (maybe == true), yet they work differently.

Is it? Why not `if (maybe != false)`?

**Comment by spydacarnage** [comment]
@canton7  - A fair question, and one that needs investigation of the underlying IL to answer.  I'm not an expert on IL codes, but it seems that both `if (maybe)` and `if (maybe == true)` produce the same code - `brfalse.s`, which, according to the reference "Transfers control to a target instruction if value is false, a null reference, or zero.", so it equates a null reference to `false`.

(not that I'm arguing against the provided reason behind the original design - I was just intrigued by you questioning something I've thought was true for a very long time!)

**Comment by mariusGundersen** [comment]
Having `if(maybe)` be equivalent to `if(maybe == true)` does not prevent using `if(!maybe)`, `if(maybe == false)`, `if(maybe != true)` or `if(maybe != false)` when that is needed. 

condition/maybe | true | false | null
-----------|------|------|-----
`if(maybe == true)` | true | false | false
`if(maybe == false)` | false | true | false
`if(maybe != true)` | false | true | true
`if(maybe != false)` | true | true | false
`if(maybe == null)` | false | false | true
`if(maybe != null)` | true | true | false
`if(maybe)` | true | false | false
`if(!maybe)` | false | true | true

**Reply by JustNrik** [reply]
I don't think it's hard to disallow `!maybe` for `bool?`.

**Comment by canton7** [comment]
@spydacarnage The IL doesn't matter - that's an implementation detail. I'm talking at the semantic level. 

You said that `if (maybe)` was definitely shorthand for `if (maybe == true)`, and I countered that this isn't true - it could just as easily be any other expression which has the same output for `bool`, but might have a different output for `bool?`. In fact, `if (thing)` is not defined if `thing` is not a `bool`, so applying it to `bool?` is nonsensical.

**Comment by xPaw** [comment]
Just to add another use case, currently with [null conditionals](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-) you have to write something like this:

```csharp
if (Thing.List?.Exists(x => x == "foo") ?? false)
```
or
```csharp
if (Thing.List !== null && Thing.List.Exists(x => x == "foo"))
```

**Reply by DavidArno** [reply]
Absolutely you do. `if` only works with boolean expressions: `if boolean expression is true then do this else do that`. A `bool?` is not a boolean expression. Nor is an `int`. So just as you have to do `if(i == 1)`, you have to do ` if (b == true)` when `b` is a `bool?`. 

Sure it makes for slightly longer syntax as in your examples. But it avoids all the pitfalls that langauges that are slapdash about boolean tests suffer from. The trade off of a bit more typing for the avoidance of nasty logic bugs is more than worth it.

**Comment by Joe4evr** [comment]
[Required reading](https://ericlippert.com/2012/03/26/null-is-not-false-part-one/).

**Reply by JustNrik** [reply]
Nobody is saying or implicating that null is false, making the whole article already pointless since it's based on a fallacy. `if` evaluates if the expression is `true` and falls into the `else` clause when the condition is not true, not when it's false.

**Reply by jnm2** [reply]
@JustNrik Your first sentence is not compatible with interpreting your comment as being in good faith.

**Reply by masonwheeler** [reply]
Umm... that person "who clearly has no idea what `if` is or how basic logic works" is one of the principal guys who wrote the compiler and created the C# language.  In a fact-based technical discussion, you're not going to persuade many people if you say things like that that clearly demonstrate you have no idea what (or in this case, *who*) you're talking about.

**Reply by JustNrik** [reply]
> 
> 
> Umm... that person "who clearly has no idea what `if` is or how basic logic works" is one of the principal guys who wrote the compiler and created the C# language. In a fact-based technical discussion, you're not going to persuade many people if you say things like that that clearly demonstrate you have no idea what (or in this case, _who_) you're talking about.

And this is another fallacy, he could be Albert Einstein, that doesn't make him absolutely right on everything he posts. The post is based on a fallacy, specifically Vacuous truth fallacy and there are more fallacies that are also applied there but cba to list them all. 

The point thing is, the post is already completely wrong from the beginning, you can't just say it's true or recommended reading just because someone with certain degree, position, or whatever said it. And the post is based in a statement that is true but it's unrelated to the issue. This blog is basically "we have A, B isn't related to A but B is true and hence A is true".

**Reply by JustNrik** [reply]
> 
> 
> @JustNrik Your first sentence is not compatible with interpreting your comment as being in good faith.

I think I exaggerated with the words I used, I will edit it, I apologize.

**Reply by HaloFour** [reply]
`null` is not `true`.  Neither is `1`, nor `""`, nor any of the countless possibilities of values that aren't a `bool`.  The team chose very intentionally that only `bool` expressions can be used for conditionals based on the real confusion and bugs that result when languages are looser in what they interpret as "truthy".  It's trivial to explicitly declare intent and I don't think anything has changed in the last 16 years that would cause that decision to be revisited.

**Reply by JustNrik** [reply]
> 
> 
> `null` is not `true`. Neither is `1`, nor `""`, nor any of the countless possibilities of values that aren't a `bool`. The team chose very intentionally that only `bool` expressions can be used for conditionals based on the real confusion and bugs that result when languages are looser in what they interpret as "truthy". It's trivial to explicitly declare intent and I don't think anything has changed in the last 16 years that would cause that decision to be revisited.

You're are using vacuous truth fallacy again, "`null` is not `true`. Neither is `1`, nor `""`, nor any of the countless possibilities of values that aren't a `bool`." yes that's true, but it's unrelated, so you are proving absolutely nothing.

```cs
bool? b = true;
```  

The value of `b` is `true`, not `null` and hence, it should be threated as `true`. If it was `null`, then it should be threated as not thruthy, which falls into the `else` clause. This, again, doesn't imply that "null is false" in any way.

**Reply by jnm2** [reply]
If that's all there was to it, then `object obj = true; if (obj) {` should work.

**Reply by elpoul** [reply]
Can we please avoid the toxicity? We're all professionals here

**Reply by jnm2** [reply]
> You're are using vacuous truth fallacy again

This is a distractingly unfriendly style of conversation, and it assumes something that I don't think you've established yet.

**Reply by JustNrik** [reply]
> 
> 
> If that's all there was to it, then `object obj = true; if (obj) {` should work.

Well that's a point, but `bool?` is supposed to be semantically equivalent to a `bool` with an additional state, which is `null`, I can understand that this decision was made for explicitness and thing, but the way everything is justified feels wrong. I would be more convinced if people said "This is intentionally disallowed for explicitness, code quality, etc etc etc" instead of throwing an unrelated "nUlL iS nOt FaLsE" premise.

**Reply by jnm2** [reply]
> I would be more convinced if people said "This is intentionally disallowed for explicitness, code quality, etc etc etc" instead of throwing an unrelated "nUlL iS nOt FaLsE" premise.

But that's exactly what everyone is saying. As Halo [said](https://github.com/dotnet/csharplang/discussions/2861#discussioncomment-124199): "Null is not false, **and treating it as either implicitly can lead to accidental logic errors.**" (Emphasis mine because your reply to his comment shows you missed the point of this sentence.)

**Reply by JustNrik** [reply]
> 
> 
> > You're are using vacuous truth fallacy again
> 
> This is a distractingly unfriendly style of conversation, and it assumes something that I don't think you've established yet.

I don't think exposing a fallacy is a "distractingly unfriendly style of conversation", and I don't do it with derogatory intentions or something. A fallacy is an argument that looks valid but it isn't, they aren't intentional all time and that's why I expose them so people stop basing their arguments in false premises. I want a proper justification of why this decision was made and not just repeat the same false premise over and over again.

Also, I think I've established it? Maybe I wasn't clear enough.

"we have A, B isn't related to A but B is true and hence A is true".

A = `if (bool?) should be interpreted as if (bool? == true)`
B = `null is not false`

B isn't related to A, and hence, A can't be proven true or false based on B.

**Reply by jnm2** [reply]
You haven't established that anyone's point so far has been "we have A, B isn't related to A but B is true and hence A is true."

> I don't think exposing a fallacy is a "distractingly unfriendly style of conversation", and I don't do it with derogatory intentions or something.

You're talking down to us and claiming to reduce what we're saying to a simple fallacy. But that's not what we were saying.

**Reply by masonwheeler** [reply]
> You're are using vacuous truth fallacy again

That's not what the fallacy means.  A [vacuous truth](https://en.wikipedia.org/wiki/Vacuous_truth) is a statement that is technically true only because there is nothing that satisfies its condition.  For example, making a statement about male calico cats would be a vacuous truth, because calico fur patterns only arise from having two X chromosomes.

`null is not false` is not a statement that applies only to an empty set of examples.

**Reply by JustNrik** [reply]
> 
> 
> > I would be more convinced if people said "This is intentionally disallowed for explicitness, code quality, etc etc etc" instead of throwing an unrelated "nUlL iS nOt FaLsE" premise.
> 
> But that's exactly what everyone is saying. As Halo [said](https://github.com/dotnet/csharplang/discussions/2861#discussioncomment-124199): "Null is not false, **and treating it as either implicitly can lead to accidental logic errors.**" (Emphasis mine because your reply to his comment shows you missed the point of this sentence.)

But no one is threating it as true or false. 

```cs
if (nullableBoolean)
{
}
else
{
}
```

1) if `nullableBoolean` is true, there's no (shouldn't be any) problem, because it's true.
2) if `nullableBoolean` is false, it should fall into the else clause, **not** because it's false, it's because it's **not true**.
3) if `nullableBoolean` is null, it should fall into the else clause, **not** because it's null, **not** because we are falsely assuming it's false, it's precisely because **null is not true** and hence it failed the thruthy test and must fall into the else clause.

**Reply by JustNrik** [reply]
> 
> 
> > You're are using vacuous truth fallacy again
> 
> That's not what the fallacy means. A [vacuous truth](https://en.wikipedia.org/wiki/Vacuous_truth) is a statement that is technically true only because there is nothing that satisfies its condition. For example, making a statement about male calico cats would be a vacuous truth, because calico fur patterns only arise from having two X chromosomes.
> 
> `null is not false` is not a statement that applies only to an empty set of examples.

"if London is in France, then the Eiffel Tower is in Bolivia"
"if null is not false, then bool?s can't be threated as true even if they are actually true"

Again, null not being false, which is completely true and I'm not against it, has nothing to do with the proposal.

**Reply by jnm2** [reply]
Branching to the `else` block fits most people's mental model of treating the value as false. As a bunch of commenters have said now, defining the condition in terms of `true` being the only special value will lead to bugs because of the assumptions people will make. `false` and `null` are bucketed together in effect if you say `true` is the only significant value, and that bucketing is the problem.

**Reply by HaloFour** [reply]
The team _very_ intentionally did not want `bool?` conditions to behave that way.

**Reply by canton7** [reply]
> if nullableBoolean is null, it should fall into the else clause, not because it's null, not because we are falsely assuming it's false, it's precisely because null is not true and hence it failed the thruthy test and must fall into the else clause.

Consider

```cs
if (!nullableBoolean)
{
} 
```

If `nullableBoolean` is `null`, then under this proposal we'll enter the `if`. `if` does a truthy test, so `!null` must be truthy. It necessarily gets treated as equivalent to `false`: you can't avoid it.

**Reply by CyrusNajmabadi** [reply]
> if evaluates if the expression is true and falls into the else clause when the condition is not true, not when it's false.

We could have done this. We considered this. We explicitly rejected this as being far too easy to introduce unintended logic errors and because of the lack of inversion symmetry. 

I don't think anyhing has changed since those original decisions that would lead do a different conclusion.

**Reply by JustNrik** [reply]
> 
> 
> > if nullableBoolean is null, it should fall into the else clause, not because it's null, not because we are falsely assuming it's false, it's precisely because null is not true and hence it failed the thruthy test and must fall into the else clause.
> 
> Consider
> 
> ```cs
> if (!nullableBoolean)
> {
> } 
> ```
> 
> If `nullableBoolean` is `null`, then under this proposal we'll enter the `if`. `if` does a truthy test, so `!null` must be truthy. It necessarily gets treated as equivalent to `false`: you can't avoid it.

It could be disallowed for `bool?`

**Reply by JustNrik** [reply]
> 
> 
> > if evaluates if the expression is true and falls into the else clause when the condition is not true, not when it's false.
> 
> We could have done this. We considered this. We explicitly rejected this as being far too easy to introduce unintended logic errors and because of the lack of inversion symmetry.
> 
> I don't think anyhing has changed since those original decisions that would lead do a different conclusion.

That reason makes much more sense than just "null is not false". I appreciate this conversation as I've clarified some doubs I had, now I see the reasons why this was rejected.

**Reply by canton7** [reply]
> It could be disallowed for bool?

At that point you've lost inversion symmetry. Why can you write:

```cs
if (condition)
{
}
else
{
}
```

but not:

```cs
if (!condition)
{
}
else
{
}
```

I don't know of any language which does that: all languages I know which allow `if` statements to take non-bool values have inversion symmetry here: `if (condition) { ... }` is equivalent to `if (!condition) { } else { ... }`.

Perhaps this is why everyone else is assuming that if something is not truthy, it must be falsy. Because the alternative is obviously a confusing idea, and no other languages have (sensibly) adopted it.

**Reply by DavidArno** [reply]
@JustNrik,
```cs
if (object)
{
}
else
{
}
```
if `object` is `true`, there's no (shouldn't be any) problem, because it's true.
... 
if `object` is `"apple"`, it should fall into the else clause, not because it's `"apple"`, not because we are falsely assuming it's false, it's precisely because `"apple"` is not true and hence it failed the thruthy test and must fall into the else clause.

Would you really want C# to treat instances of `Object` in this way? An instance of `Object` could be a boolean and it could be true, just as an instance of `Nullable<bool>` could be a boolean and it could be true. But just because it could be a boolean, doesn't mean it is a boolean nor should it be allowed in place of a boolean expression.

**Comment by The-Futurist** [comment]
```cs
        static void Main(string[] args)
        {
            bool? nullable_bool = null;

            if (False(nullable_bool))
                ; // whatever
        }

        static bool True (bool? Arg)
        {
            if (Arg.HasValue && Arg.Value)
                return true;

            return false;
        }

        static bool False (bool? Arg)
        {
            return !True(Arg);
        }

```

**Reply by jnm2** [reply]
This `True(x)` could be `x is true`, and this `False(x)` could be `x is false or null` or `x is not true`.

**Reply by canton7** [reply]
Or just `x == true` and `x != true`

**Reply by jnm2** [reply]
Or `x ?? false` and `x ?? true` 😁 I guess it depends what @Korporal was going for with this.
